Sources/SourceKitLSP/Swift/DocumentFormatting.swift

```diff
@@ -139,26 +139,58 @@ private func edits(from original: DocumentSnapshot, to edited: String) -> [TextE
 extension SwiftLanguageService {
   package func documentFormatting(_ req: DocumentFormattingRequest) async throws -> [TextEdit]? {
     return try await format(
+      snapshot: documentManager.latestSnapshot(req.textDocument.uri),
       textDocument: req.textDocument,
       options: req.options
     )
   }
 
   package func documentRangeFormatting(_ req: DocumentRangeFormattingRequest) async throws -> [TextEdit]? {
     return try await format(
+      snapshot: documentManager.latestSnapshot(req.textDocument.uri),
       textDocument: req.textDocument,
       options: req.options,
       range: req.range
     )
   }
 
+  package func documentOnTypeFormatting(_ req: DocumentOnTypeFormattingRequest) async throws -> [TextEdit]? {
+    guard let server = self.sourceKitLSPServer else {
+      return nil
+    }
+    let snapshot = try documentManager.latestSnapshot(req.textDocument.uri)
+    let capabilities = await server.serverCapabilities(
+      for: capabilityRegistry.clientCapabilities,
+      registry: capabilityRegistry
+    )
+    
+    guard let documentOnTypeFormattingProvider = capabilities.documentOnTypeFormattingProvider,
+      documentOnTypeFormattingProvider.firstTriggerCharacter == req.ch
+        || documentOnTypeFormattingProvider.moreTriggerCharacter?.contains(req.ch) == true,
+      let line = snapshot.lineTable.line(at: req.position.line),
+      /// No need to go through whitespace checking if the trigger is not a newline
+      !req.ch.isNewline || !line.allSatisfy(\.isWhitespace)
+    else {
+      return nil
+    }
+
+    let lineStart = Position(line: req.position.line, utf16index: 0)
+    let nextLineStart = Position(line: req.position.line + 1, utf16index: 0)
```

Comment:
What if this is the last line and there’s no trailing newline? Then we should use the end index of `req.position.line` to not go out of bounds.

Reply:
Does it matter considering the range is exclusive so the next line is never actually operated on?

Hmm i see how this could wrong though, I'll need to adjust the code.

If the next line doesn't exist, `snapshot` functions will just return the `endIndex` of the content.
That means `upperBound` could be the last `utf8Offset` in the same line. And we're doing `upperBound - 1` because Swift's `Range.upperBound` is exclusive, but in this case we'll be ignoring the last character of the last line which will be a bug.

Reply:
The most important thing that I thought of was that the conversion functions would log a fault despite there not being an actual issue.

Reply:
I changed the code to find the positions more-strictly:
https://github.com/swiftlang/sourcekit-lsp/pull/1815/commits/51fab20c8c23adf18e54e219f3b052166f2ac3ab